### PR TITLE
PhysicalExpr Orderings with Range Information

### PR DESCRIPTION
**Pull request by berkaysynnada**
## Which issue does this PR close?

Closes #.

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/enforce_distribution.rs [hunk 1]
```diff
@@ -3572,7 +3572,11 @@ pub(crate) mod tests {
             expr: col("c", &schema).unwrap(),
             options: SortOptions::default(),
         }];
-        let alias = vec![("a".to_string(), "a".to_string())];
+        let alias = vec![
+            ("a".to_string(), "a".to_string()),
+            ("b".to_string(), "b".to_string()),
+            ("c".to_string(), "c".to_string()),
+        ];
         let plan = sort_preserving_merge_exec(
```
Comment (berkaysynnada) [inline datafusion/core/src/physical_optimizer/enforce_distribution.rs:3580]: A test bug shows up, fixed here.

File: datafusion/common/src/scalar/mod.rs [hunk 2]
```diff
@@ -991,6 +991,59 @@ impl ScalarValue {
         })
     }
 
+    /// Create a null value in the given type.
+    pub fn new_null(datatype: &DataType) -> Result<ScalarValue> {
```
Comment (mustafasrepo) [inline datafusion/common/src/scalar/mod.rs]: You can use `try_from` for this work.

File: datafusion/functions/src/datetime/date_bin.rs [hunk 3]
```diff
@@ -146,8 +146,8 @@ impl ScalarUDFImpl for DateBinFunc {
         }
     }
 
-    fn monotonicity(&self) -> Result<Option<FuncMonotonicity>> {
-        Ok(Some(vec![None, Some(true)]))
+    fn monotonicity(&self, input: &[ExprProperties]) -> Result<SortProperties> {
+        Ok(input[1].sort_properties)
```
Comment (mustafasrepo) [inline datafusion/functions/src/datetime/date_bin.rs]: You can add comment here.

File: datafusion/functions/src/datetime/date_trunc.rs [hunk 4]
```diff
@@ -205,8 +205,8 @@ impl ScalarUDFImpl for DateTruncFunc {
         &self.aliases
     }
 
-    fn monotonicity(&self) -> Result<Option<FuncMonotonicity>> {
-        Ok(Some(vec![None, Some(true)]))
+    fn monotonicity(&self, input: &[ExprProperties]) -> Result<SortProperties> {
+        Ok(input[1].sort_properties)
```
Comment (mustafasrepo) [inline datafusion/functions/src/datetime/date_trunc.rs]: You can add comment here

File: datafusion/functions/src/math/pi.rs [hunk 5]
```diff
@@ -70,7 +71,7 @@ impl ScalarUDFImpl for PiFunc {
         ))))
     }
 
-    fn monotonicity(&self) -> Result<Option<FuncMonotonicity>> {
-        Ok(Some(vec![Some(true)]))
+    fn monotonicity(&self, input: &[ExprProperties]) -> Result<SortProperties> {
+        Ok(input[0].sort_properties)
```
Comment (mustafasrepo) [inline datafusion/functions/src/math/pi.rs]: you can add comment here

Comment (mustafasrepo) [reply datafusion/functions/src/math/pi.rs]: It should return singleton

File: datafusion/functions/src/math/round.rs [hunk 6]
```diff
@@ -80,8 +81,8 @@ impl ScalarUDFImpl for RoundFunc {
         make_scalar_function(round, vec![])(args)
     }
 
-    fn monotonicity(&self) -> Result<Option<FuncMonotonicity>> {
-        Ok(Some(vec![Some(true)]))
+    fn monotonicity(&self, input: &[ExprProperties]) -> Result<SortProperties> {
+        Ok(input[0].sort_properties)
```
Comment (mustafasrepo) [inline datafusion/functions/src/math/round.rs]: You can add comment here

File: datafusion/functions/src/math/trunc.rs [hunk 7]
```diff
@@ -86,8 +87,8 @@ impl ScalarUDFImpl for TruncFunc {
         make_scalar_function(trunc, vec![])(args)
     }
 
-    fn monotonicity(&self) -> Result<Option<FuncMonotonicity>> {
-        Ok(Some(vec![Some(true)]))
+    fn monotonicity(&self, input: &[ExprProperties]) -> Result<SortProperties> {
+        Ok(input[0].sort_properties)
```
Comment (mustafasrepo) [inline datafusion/functions/src/math/trunc.rs]: Can add comment

File: datafusion/physical-expr/src/utils/mod.rs [hunk 8]
```diff
@@ -309,8 +308,8 @@ pub(crate) mod tests {
             }
         }
 
-        fn monotonicity(&self) -> Result<Option<FuncMonotonicity>> {
-            Ok(Some(vec![Some(true)]))
+        fn monotonicity(&self, input: &[ExprProperties]) -> Result<SortProperties> {
+            Ok(input[0].sort_properties)
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/utils/mod.rs:312]: Add comment

File: datafusion/sqllogictest/test_files/order.slt [hunk 9]
```diff
@@ -955,3 +955,81 @@ drop table foo;
...
+04)------TableScan: ordered_table projection=[c]
+physical_plan
+01)GlobalLimitExec: skip=0, fetch=5
+02)--SortPreservingMergeExec: [c_str@0 ASC NULLS LAST], fetch=5
+03)----SortExec: TopK(fetch=5), expr=[c_str@0 ASC NULLS LAST], preserve_partitioning=[true]
+04)------ProjectionExec: expr=[CAST(c@0 AS Utf8) as c_str]
+05)--------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
+06)----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/window_2.csv]]}, projection=[c], output_ordering=[c@0 ASC NULLS LAST], has_header=true
+
+query I
```
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/order.slt:1006]: Add comment. here numeric type preserves ordering

**Review by mustafasrepo — APPROVED** [review]
LGTM!

**Comment by ozankabak** [comment]
Merged upstream, closing.

**Pull request closed without merging**